internal/pkg/agent/cmd/run.go

```diff
@@ -557,3 +561,43 @@ func setupMetrics(
 func isProcessStatsEnabled(cfg *monitoringCfg.MonitoringConfig) bool {
 	return cfg != nil && cfg.HTTP.Enabled
 }
+
+// handleUpgrade checks if agent is being run as part of an
+// ongoing upgrade operation, i.e. being re-exec'd and performs
+// any upgrade-specific work, if needed.
+func handleUpgrade() error {
+	upgradeMarker, err := upgrade.LoadMarker()
+	if err != nil {
+		return fmt.Errorf("unable to load upgrade marker to check if Agent is being upgraded: %w", err)
+	}
+
+	if upgradeMarker == nil {
+		// We're not being upgraded. Nothing more to do.
+		return nil
+	}
+
+	// In v8.8.0, we introduced a new installation marker file to indicate that
+	// an Agent was running as installed. When an installed Agent that's older
+	// than v8.8.0 is upgraded, this installation marker file is not present.
+	// So, in such cases, we need to create it manually post-upgrade.
+	// Otherwise, the upgrade will be unsuccessful (see
+	// https://github.com/elastic/elastic-agent/issues/2645).
+
+	installMarkerVersion := semver.MustParse("8.8.0")
+	prevVersion, err := semver.NewVersion(upgradeMarker.PrevVersion)
```

Comment:
Probably worth checking to see if this `PrevVersion` field always exists in every version of the upgrade marker file that can be created. The 7.17.0 release is a good place to start checking.

Does 8.8.0 always create the installation marker? If it does then the absence of the installation marker implies that this is an agent version prior to 8.8.0 installed in the default location since that was the only option. Is that possibly safer since it doesn't depend on the format of the upgrade marker file?

Reply:
I have no issue with approach as long as we can prove it will always work.

Reply:
> Does 8.8.0 always create the installation marker? If it does then the absence of the installation marker implies that this is an agent version prior to 8.8.0 installed in the default location since that was the only option. Is that possibly safer since it doesn't depend on the format of the upgrade marker file?

Yes, we always create the installation marker in 8.8.0 when Agent is installed using `elastic-agent install`.  That said, I also like the idea of not depending on the _format_ of the upgrade marker file to determine the pre-upgrade version. 

I do think we need to check for the _existence_ of the upgrade marker file, though, to know that we're in the midst of an upgrade.  Otherwise, we'll end up creating the installation marker even if a user simply ran `elastic-agent run`  or more likely, in the Docker use case (probably harmless).

Reply:
Yes, checking for the existence of the file should be fine. Looking inside it is more likely to have unexpected results. I'm not sure what guarantees we have about its format so that likely means we have none.

Reply:
The format is guaranteed by the mapping of the [`UpdateMarker` struct](https://github.com/elastic/elastic-agent/blob/4cfce531d3b9be8513a130bfe1c6f27fa9adddd0/internal/pkg/agent/application/upgrade/step_mark.go#L26L40)
It's not a file that a user is supposed to touch...